### PR TITLE
Extend PVCAction itemblock plugin to support grouping PVCs under VGS label key

### PR DESCRIPTION
**Pull request by shubham-pampattiwar**
Thank you for contributing to Velero!

# Please add a summary of your change

- Adds support in GetRelatedItems for grouping PVCs using the VGS label key provided.
- Related PVCs in the same VGS group are included in the ItemBlock to ensure atomic snapshot coverage.

# Does your change fix a particular issue?

Fixes https://github.com/vmware-tanzu/velero/issues/8935

Related to sub-task of https://github.com/vmware-tanzu/velero/issues/8865

~~Depends on merging https://github.com/vmware-tanzu/velero/pull/8938~~

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [x] [Created a changelog file (`make new-changelog`)](https://velero.io/docs/main/code-standards/#adding-a-changelog) or comment `/kind changelog-not-required` on this PR.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

File: pkg/itemblock/actions/pvc_action.go [hunk 1]
```diff
@@ -105,9 +105,61 @@
...
+		a.log.Info("No VolumeGroupSnapshotLabelKey provided in backup spec; skipping PVC grouping")
+		return nil, nil
+	}
+
+	groupID, ok := pvc.Labels[vgsLabelKey]
+	if !ok {
+		// PVC does not belong to any VGS group.
+		return nil, nil
+	}
+
```
Comment (kaovilai) [inline pkg/itemblock/actions/pvc_action.go:139]: **Enhancement Suggestion:** Consider adding validation for empty group values to prevent potential issues:

```go
groupID, ok := pvc.Labels[vgsLabelKey]
if !ok || groupID == "" {
    // PVC does not belong to any VGS group or has empty group value
    return nil, nil
}
```

This prevents processing PVCs that have the label key but with an empty value, which could cause unexpected behavior.

File: pkg/itemblock/actions/pvc_action.go [hunk 2]
```diff
@@ -105,9 +105,61 @@
...
+	pvcList := new(corev1api.PersistentVolumeClaimList)
+	if err := a.crClient.List(
+		ctx,
+		pvcList,
+		crclient.InNamespace(pvc.Namespace),
+		crclient.MatchingLabels{vgsLabelKey: groupID},
+	); err != nil {
+		return nil, errors.Wrap(err, "failed to list PVCs for VGS grouping")
+	}
+
```
Comment (kaovilai) [inline pkg/itemblock/actions/pvc_action.go:149]: **Enhancement Suggestion:** Add more context to the error message for better debugging:

```go
return nil, errors.Wrapf(err, "failed to list PVCs for VGS grouping with label %s=%s in namespace %s", vgsLabelKey, groupID, pvc.Namespace)
```

This provides more detailed information when troubleshooting VGS grouping issues.

File: pkg/itemblock/actions/pvc_action.go [hunk 3]
```diff
@@ -105,9 +105,61 @@
...
+	for _, groupPVC := range pvcList.Items {
+		if groupPVC.Name == pvc.Name {
```
Comment (kaovilai) [inline pkg/itemblock/actions/pvc_action.go:151]: **Performance Enhancement:** Consider adding an early return optimization after the List call:

```go
if len(pvcList.Items) <= 1 {
    // Only the current PVC exists in this group
    return nil, nil
}
```

This avoids unnecessary processing when no grouping is actually needed. Also consider pre-allocating the slice capacity:

```go
related := make([]velero.ResourceIdentifier, 0, len(pvcList.Items)-1)
```

**Comment by kaovilai** [comment]
## Code Review Summary ✅

Great work on implementing VGS support in the PVC ItemBlockAction plugin! This is a solid implementation that addresses issue #8935 requirements.

### **Overall Assessment:**
- ✅ **Clean implementation** with proper separation of concerns
- ✅ **Excellent test coverage** covering edge cases
- ✅ **Good error handling** and logging
- ✅ **Bug fix included** (method name correction)
- ✅ **Efficient Kubernetes client usage**

### **Key Strengths:**
1. **Namespace scoping**: Correctly limits PVC search to the same namespace
2. **Label-based grouping**: Uses efficient `MatchingLabels` client filtering
3. **Self-exclusion**: Properly excludes the current PVC from grouped results
4. **Graceful degradation**: Handles missing VGS configuration cleanly

### **Minor Enhancement Opportunities** (within PR scope):
I've added specific code review comments for:
- Empty group value validation
- Enhanced error context for debugging
- Performance optimizations (early returns, pre-allocated slices)

### **Test Coverage:**
The test suite is comprehensive. Consider adding one more edge case for empty group values to complete the coverage.

### **Documentation:**
As mentioned, documentation is covered in the parent issue #8865, so this PR focuses correctly on the implementation scope.

**Recommendation:** ✅ **Approve** once the minor suggestions are considered. This feature will be valuable for atomic backup operations with Volume Group Snapshots.

Looking forward to seeing this land! 🚀

File: pkg/itemblock/actions/pvc_action.go (hunk 1 above)
Comment (shubham-pampattiwar) [reply pkg/itemblock/actions/pvc_action.go:140]: Done, updated the PR !

File: pkg/itemblock/actions/pvc_action.go (hunk 2 above)
Comment (shubham-pampattiwar) [reply pkg/itemblock/actions/pvc_action.go:155]: Done, updated the PR !

File: pkg/itemblock/actions/pvc_action.go (hunk 3 above)
Comment (shubham-pampattiwar) [reply pkg/itemblock/actions/pvc_action.go:157]: Done, updated the PR !

**Comment by shubham-pampattiwar** [comment]
@sseago Addressed https://github.com/vmware-tanzu/velero/pull/8938#pullrequestreview-2881635323 in this PR !

File: pkg/itemblock/actions/pvc_action.go [hunk 4]
```diff
@@ -105,9 +105,67 @@
...
+	pvcList := new(corev1api.PersistentVolumeClaimList)
+	if err := a.crClient.List(
+		ctx,
+		pvcList,
+		crclient.InNamespace(pvc.Namespace),
+		crclient.MatchingLabels{vgsLabelKey: groupID},
+	); err != nil {
+		return nil, errors.Wrapf(err, "failed to list PVCs for VGS grouping with label %s=%s in namespace %s", vgsLabelKey, groupID, pvc.Namespace)
+	}
+
```
Comment (kaovilai) [inline pkg/itemblock/actions/pvc_action.go:150]: **Performance Optimization Suggestion:** Consider excluding the current PVC at the query level rather than filtering it out after fetching all results. This would be more efficient, especially for large volume groups:

```go
// More efficient query that excludes current PVC at the API level
if err := a.crClient.List(
    ctx,
    pvcList,
    crclient.InNamespace(pvc.Namespace),
    crclient.MatchingLabels{vgsLabelKey: groupID},
    crclient.MatchingFields{"metadata.name!=": pvc.Name}, // If supported by the client
); err != nil {
```

Note: This optimization depends on whether the controller-runtime client supports field selectors with inequality operators. If not supported, the current implementation is fine, but it might be worth investigating for future optimization.

**Review by kaovilai — DISMISSED** [review]
LGTM!

File: pkg/itemblock/actions/pvc_action.go [hunk 5]
```diff
@@ -105,9 +105,67 @@ func (a *PVCAction) GetRelatedItems(item runtime.Unstructured, backup *v1.Backup
...
+	return "PVCItemBlockAction"
+}
+
+// getGroupedPVCs returns other PVCs in the same group based on the VGS label key in the Backup spec.
+func (a *PVCAction) getGroupedPVCs(ctx context.Context, pvc *corev1api.PersistentVolumeClaim, backup *v1.Backup) ([]velero.ResourceIdentifier, error) {
+	var related []velero.ResourceIdentifier
+
+	vgsLabelKey := backup.Spec.VolumeGroupSnapshotLabelKey
+	if vgsLabelKey == "" {
+		a.log.Info("No VolumeGroupSnapshotLabelKey provided in backup spec; skipping PVC grouping")
```
Comment (sseago) [inline pkg/itemblock/actions/pvc_action.go]: Should this be a debug log? I'm not sure we need a new log msg for every PVC backed up here if we're not using VGS.

Comment (sseago) [reply pkg/itemblock/actions/pvc_action.go]: Also, this should basically never trigger, right? Since velero will fill in the default one if one isn't specified? Still need the check, but it shouldn't actually get to this line if things are working right.

File: pkg/itemblock/actions/pvc_action.go [hunk 6]
```diff
@@ -105,9 +105,67 @@ func (a *PVCAction) GetRelatedItems(item runtime.Unstructured, backup *v1.Backup
...
+		return nil, nil
+	}
+
+	groupID, ok := pvc.Labels[vgsLabelKey]
+	if !ok || groupID == "" {
+		// PVC does not belong to any VGS group or groupID has empty value
+		a.log.Info("PVC does not belong to any PVC group or group label value is empty; skipping PVC grouping")
```
Comment (sseago) [inline pkg/itemblock/actions/pvc_action.go]: Same comment as above about debug log.

File: pkg/itemblock/actions/pvc_action.go (hunk 5 above)
Comment (shubham-pampattiwar) [reply pkg/itemblock/actions/pvc_action.go]: Done, updated the PR !

File: pkg/itemblock/actions/pvc_action.go (hunk 6 above)
Comment (shubham-pampattiwar) [reply pkg/itemblock/actions/pvc_action.go]: Done, updated the PR !

File: pkg/controller/backup_controller_test.go [hunk 7]
```diff
@@ -484,7 +484,7 @@ func TestPrepareBackupRequest_SetsVGSLabelKey(t *testing.T) {
 	require.NoError(t, err)
 	now = now.Local()
 
-	defaultVGSLabelKey := "velero.io/volume-group-snapshot"
+	defaultVGSLabelKey := "velero.io/volume-group"
```
Comment (kaovilai) [inline pkg/controller/backup_controller_test.go]: These are being defined multiple places. Maybe use https://github.com/vmware-tanzu/velero/blob/d2c6b6bc3e81e749d2e2b0c1ff2872ab18f2523e/pkg/apis/velero/v1/labels_annotations.go ?

Comment (shubham-pampattiwar) [reply pkg/controller/backup_controller_test.go]: Done

**Review by sseago — APPROVED** [review]

**Comment by shubham-pampattiwar** [comment]
@Lyndon-Li @blackpiglet @kaovilai PR review request !

**Review by Lyndon-Li — APPROVED** [review]
